open-craft/sprints
# [BB-3175] Add asynchronous sprint automation

This adds Celery tasks and utils for asynchronous sprint automation.

## Prerequisites
1. Set `.env` (you can find them [here](https://vault.opencraft.com:8200/ui/vault/secrets/secret/show/core/Sprints%20Development%20Envs)). There were some recent changes.
1. In `sprints/dashboard/automation.py:notify_about_injection`, add `print(issue, message)` above `if not settings.DEBUG`.
1. In `sprints/dashboard/tasks.py:check_tickets_ready_for_sprint_task`, add `print(f"#{cell}: {emails}, {message}")` above `if not settings.DEBUG`.
1. In `sprints/dashboard/tasks.py:ping_overcommitted_users_task`, add `print(f"#{cell}: {emails}, {message}")` above `if not settings.DEBUG`.

## Testing instructions 
1. Start the backend with `docker-compose -f local.yml up --build` (this is important because we're changing dependencies here).
1. Run migrations with `docker-compose -f local.yml exec django python manage.py migrate`.
1. Create superuser with `docker-compose -f local.yml exec django python manage.py createsuperuser`
1. Open Django shell (in a separate terminal) with `docker-compose -f local.yml exec django python manage.py shell_plus` and input the following:
    ```python
    from sprints.dashboard.automation import *
    from sprints.dashboard.tasks import *
    ```
1. Run `move_out_injections_task()` and check that the outputted comments contain [BB-3630](https://tasks.opencraft.com/browse/BB-3630).
1. Go to [BB-3630](https://tasks.opencraft.com/browse/BB-3630) and add a `injection-accepted` label there.
1. Run `move_out_injections_task()` and check that the outputted comments do not contain [BB-3630](https://tasks.opencraft.com/browse/BB-3630) anymore.
1. Run `check_tickets_ready_for_sprint_task()` and check that the output makes sense. Notice that there is a special case in [BB-3668](https://tasks.opencraft.com/browse/BB-3668), as the reporter is not a member of any cell.
1. Run `ping_overcommitted_users_task()` and check that mentioned users are overcommitted (on sprints.opencraft.com).
1. Run and check the following:
    ```python
    conn = CustomJira(
        server=settings.JIRA_SERVER,
        basic_auth=(settings.JIRA_USERNAME, settings.JIRA_PASSWORD),
        options={'agile_rest_path': GreenHopperResource.AGILE_BASE_REST_PATH}
    )
    issue = conn.issue('BB-3636')
    flag_issue(conn, issue)
    # Check that the ticket is flagged now.
    unflag_issue(conn, issue)
    # Check that the ticket is unflagged now.
    ```
1. Run `schedule_sprint_tasks_task()`, then go to [Django admin](http://localhost:8000/admin/django_celery_beat/periodictask/) and check that all four tasks have been scheduled properly for the current sprint.
